### PR TITLE
[13.0][IMP] intrastat*: brexit support

### PR DESCRIPTION
**Pull request by victoralmau**
Backport from 14.0: https://github.com/OCA/intrastat-extrastat/pull/163

Please @pedrobaeza can you review it?

@Tecnativa TT37486

### CODE REVIEW

**Comment by luc-demeyer** [comment]
@victoralmau 
I had placed this action on my todo list but now it's done already, top !.

But there is something wrong with this PR:
Stack trace when I test it;:

  File "/opt/odoo13/OCA/intrastat-extrastat/intrastat_product/models/intrastat_product_declaration.py", line 762, in action_gather
    lines = self._gather_invoices()
  File "/opt/odoo13/OCA/intrastat-extrastat/intrastat_product/models/intrastat_product_declaration.py", line 625, in _gather_invoices
    partner_country = self._get_partner_country(inv_line, eu_countries)
  File "/opt/odoo13/OCA/intrastat-extrastat/intrastat_product/models/intrastat_product_declaration.py", line 288, in _get_partner_country
    self._format_line_note(inv_line, line_notes)
TypeError: _format_line_note() missing 1 required positional argument: 'line_notes'

The signature of the _format_line_note in 13.0 is:
def _format_line_note(self, line, line_nbr, line_notes):
In 14.0 this signature has changed to:
def _format_line_note(self, line, notedict, line_notes):

Backporting also the unit tests that we have in 14.0 would be great so that we detect such issues automatically.

File: intrastat_product/models/intrastat_product_declaration.py [hunk 1]
```diff
@@ -239,15 +240,71 @@ def _account_config_warning(self, msg):
...
+                    "or on the delivery address (partner '%s'). "
+                )
+                % (
+                    inv.partner_id.display_name,
+                    inv.partner_shipping_id
+                    and inv.partner_shipping_id.display_name
+                    or "-",
+                )
+            ]
+            self._format_line_note(inv_line, line_notes)
```
Comment (luc-demeyer) [inline intrastat_product/models/intrastat_product_declaration.py]: wrong signature on _format_line_notes()

File: intrastat_product/models/intrastat_product_declaration.py [hunk 2]
```diff
@@ -239,15 +240,71 @@ def _account_config_warning(self, msg):
...
+        else:
+            if country not in eu_countries and country.code != "GB":
+                line_notes = [
+                    _(
+                        "On invoice '%s', the source/destination country "
+                        "is '%s' which is not part of the European Union."
+                    )
+                    % (inv.name, country.name)
+                ]
+                self._format_line_note(inv_line, line_notes)
```
Comment (luc-demeyer) [inline intrastat_product/models/intrastat_product_declaration.py]: wrong signature on _format_line_notes()

File: intrastat_product/models/intrastat_product_declaration.py [hunk 3]
```diff
@@ -239,15 +240,71 @@ def _account_config_warning(self, msg):
...
+                        "set the VAT number of the partner '%s' in Odoo with "
+                        "its new VAT number starting with 'XI' following Brexit."
+                    )
+                    % (
+                        inv.name,
+                        inv.fiscal_position_id.display_name,
+                        inv.commercial_partner_id.display_name,
+                    )
+                ]
+                self._format_line_note(inv_line, line_notes)
```
Comment (luc-demeyer) [inline intrastat_product/models/intrastat_product_declaration.py]: wrong signature on _format_line_notes()

File: intrastat_product/models/intrastat_product_declaration.py [hunk 4]
```diff
@@ -239,15 +240,71 @@ def _account_config_warning(self, msg):
...
+                        "its new VAT number starting with 'XI' following Brexit."
+                    )
+                    % (
+                        inv.name,
+                        inv.fiscal_position_id.display_name,
+                        vat,
+                        inv.commercial_partner_id.display_name,
+                    )
+                ]
+                self._format_line_note(inv_line, line_notes)
```
Comment (luc-demeyer) [inline intrastat_product/models/intrastat_product_declaration.py]: wrong signature on _format_line_notes()

**Review by luc-demeyer — CHANGES_REQUESTED** [review]

**Comment by victoralmau** [comment]
Changes done and tests solved.

File: intrastat_product/models/account_move.py [hunk 5]
```diff
@@ -158,11 +171,23 @@ class AccountMoveIntrastatLine(models.Model):
...
+    # adapted accordingly
     product_origin_country_id = fields.Many2one(
         comodel_name="res.country",
         string="Country of Origin",
         help="Country of origin of the product i.e. product " "'made in ____'.",
     )
+    product_origin_country_code = fields.Char(
+        string="Country of Origin of the Product",
+        size=2,
+        required=True,
```
Comment (pedrobaeza) [inline intrastat_product/models/account_move.py:185]: What happen with existing lines not having this filled?

File: intrastat_product/models/account_move.py [hunk 6]
```diff
@@ -174,3 +199,18 @@ def _onchange_move_id(self):
...
+    def create(self, vals):
+        self._format_vals(vals)
+        return super().create(vals)
+
+    def write(self, vals):
+        self._format_vals(vals)
+        return super().write(vals)
+
+    def _format_vals(self, vals):
+        if "product_origin_country_code" in vals:
```
Comment (pedrobaeza) [inline intrastat_product/models/account_move.py:213]: Why is this needed?

File: intrastat_product/models/intrastat_product_declaration.py [hunk 7]
```diff
@@ -925,6 +995,15 @@ class IntrastatProductComputationLine(models.Model):
         help="Country of Origin/Destination",
         domain=[("intrastat", "=", True)],
     )
+    src_dest_country_code = fields.Char(
+        string="Country Code",
+        compute="_compute_src_dest_country_code",
+        store=True,
+        required=True,
```
Comment (pedrobaeza) [inline intrastat_product/models/intrastat_product_declaration.py:1002]: The same for this.

File: intrastat_product/models/intrastat_product_declaration.py [hunk 8]
```diff
@@ -962,14 +1041,36 @@ class IntrastatProductComputationLine(models.Model):
...
+    # adapted accordingly
     product_origin_country_id = fields.Many2one(
         "res.country",
         string="Country of Origin of the Product",
         help="Country of origin of the product i.e. product 'made in ____'",
     )
+    product_origin_country_code = fields.Char(
+        string="Country of Origin of the Product",
+        size=2,
+        required=True,
```
Comment (pedrobaeza) [inline intrastat_product/models/intrastat_product_declaration.py:1055]: I don't continue saying, but there's a lot of new required fields with the same problem.

File: intrastat_product/tests/common.py [hunk 9]
```diff
@@ -0,0 +1,125 @@
+# Copyright 2021 ACSONE SA/NV
```
Comment (pedrobaeza) [inline intrastat_product/tests/common.py:1]: Is this copyright correct?

**Review by pedrobaeza — COMMENTED** [review]
Squash commits

File: intrastat_product/models/account_move.py (hunk 5 above)
Comment (victoralmau) [reply intrastat_product/models/account_move.py:185]: I think it is a strange behavior, but it is what was added in v14. In my opinion, this should be a compute store field. What do you think about this?

Comment (pedrobaeza) [reply intrastat_product/models/account_move.py:185]: We should keep the changes at minimum to comply with the Brexit requirement

Comment (victoralmau) [reply intrastat_product/models/account_move.py:185]: The `product_origin_country_code` field was added by the v14 PR to add the Brexit requirement (it is not country-related because it depends on the state).

You can change that field and make it a compute store, although it would be necessary to add a `product_origin_state_id` field as well (this would avoid overwriting the create() and write() of the account.move).
In any case, migration scripts will be necessary. Which way do you think is better?

**Comment by luc-demeyer** [comment]
Product Origin country code is XI, XU for brexit support.
This is a required field for the declaration.

**Comment by victoralmau** [comment]
> Product Origin country code is XI, XU for brexit support. This is a required field for the declaration.

Ping @pedrobaeza

**Comment by pedrobaeza** [comment]
I don't remember what I have to check here, but please go on in the best way you think. There is also a "Requested changes" from @luc-demeyer. Is it attended?

**Comment by victoralmau** [comment]
> I don't remember what I have to check here, but please go on in the best way you think. There is also a "Requested changes" from @luc-demeyer. Is it attended?

There was really nothing to validate (you asked if everything added since v14 was necessary). The changes proposed by @luc-demeyer have already been made so it would be ready to merge.

**Comment by pedrobaeza** [comment]
OK, let's merge it then:

/ocabot merge major

**Comment by JordiBForgeFlow** [comment]
@victoralmau @pedrobaeza buddies, where is the method '_compute_src_dest_country_code' https://github.com/OCA/intrastat-extrastat/pull/179/files#diff-74e1197466e3c9a2c08b006d373e86cca69ac389f34c9e4ed38dc24b84018d80R1000 defined??

**Comment by JordiMForgeFlow** [comment]
> @victoralmau @pedrobaeza buddies, where is the method '_compute_src_dest_country_code' https://github.com/OCA/intrastat-extrastat/pull/179/files#diff-74e1197466e3c9a2c08b006d373e86cca69ac389f34c9e4ed38dc24b84018d80R1000 defined??

In V14 I don't see it being a computed field
https://github.com/OCA/intrastat-extrastat/blob/5b436213364228fbbbba94287f5736d7e85271dc/intrastat_product/models/intrastat_product_declaration.py#L1086

I think we need to BP this fix: https://github.com/OCA/intrastat-extrastat/commit/3d4e31125eb75ebe5f420f7ff0c8ed0e949c2888

**Comment by JordiBForgeFlow** [comment]
@victoralmau please backport https://github.com/OCA/intrastat-extrastat/pull/174 ASAP

**Comment by JordiBForgeFlow** [comment]
@victoralmau please make sure that you review and backport all commits from 14.0

**Comment by victoralmau** [comment]
The missing change is now merged https://github.com/OCA/intrastat-extrastat/pull/188

**Comment by JordiBForgeFlow** [comment]
@victoralmau thanks!!
